core/cloudflow-cr-generator/src/main/scala/cloudflow/cr/Generator.scala

```diff
@@ -0,0 +1,137 @@
+      dockerImages: String => String): String = {
+    val scanConfig =
+      DescriptorExtractor.ScanConfiguration(projectId = projectId, classpathUrls = classpath.map(new URL(_)).toArray)
+
+    val extractedStreamlets = DescriptorExtractor.scan(scanConfig)
+
+    val streamletDescriptors = extractedStreamlets.map {
+      case (_, configDescriptor) =>
+        configDescriptor
+          .root()
+          .render(ConfigRenderOptions.concise())
+          .parseJson
+          .convertTo[cloudflow.blueprint.StreamletDescriptor]
+    }
+
+    val blueprintConfig = ConfigFactory.parseString(blueprintStr).resolve()
+
+    val blueprint = Blueprint.parseConfig(blueprintConfig, streamletDescriptors.toVector)
+    if (blueprint.problems.isEmpty) {
+      val appDescriptor = {
+        val appVersion = version
+        val agentPathsMap = Map("prometheus" -> "/prometheus/jmx_prometheus_javaagent.jar")
+        val libraryVersion = cloudflowVersion
+
+        ApplicationDescriptor(
+          appId = projectId,
+          appVersion = version,
+          (k) => dockerImages(k),
+          blueprint.verified.right.get,
+          agentPathsMap,
+          cloudflowVersion)
+      }
+
+      import cloudflow.blueprint.deployment.CloudflowCRFormat._
+      val cloudflowCr = makeCR(appDescriptor, version)
+      cloudflowCr.toJson.compactPrint
+    } else {
+      throw new Exception(s"Found errors: ${blueprint.problems.mkString}")
+    }
+  }
+
+  case class Config(
+      name: String = null,
+      version: String = null,
+      blueprint: File = null,
+      classpath: File = null,
+      images: Map[String, String] = Map())
+
+  private lazy val builder = OParser.builder[Config]
+  private lazy val parser = {
+    import builder._
+    OParser.sequence(
+      programName("cloudflow-cr-generator"),
+      head("cloudflow-cr-generator", cloudflowVersion),
+      opt[String]('n', "name")
+        .action((x, c) => c.copy(name = x))
+        .text("Name of the Cloudflow Application"),
+      opt[String]('v', "version")
+        .action((x, c) => c.copy(version = x))
+        .text("Version of the Cloudflow Application"),
+      opt[File]('b', "blueprint")
+        .action((x, c) => c.copy(blueprint = x))
+        .text("Blueprint of the Cloudflow Application"),
+      opt[File]('c', "classpath")
+        .action((x, c) => c.copy(classpath = x))
+        .text("Classpath of the Cloudflow Application"),
+      opt[Map[String, String]]('i', "images")
+        .action((x, c) => c.copy(images = c.images ++ x))
+        .text("Docker images of the Cloudflow Application"),
+      checkConfig(
+        c =>
+          if (c.name == null || c.version == null || c.blueprint == null || c.classpath == null)
+            failure("All the options should be filled in")
+          else success))
+  }
+
+  def main(args: Array[String]): Unit = {
+
+    OParser.parse(parser, args, Config()) match {
+      case Some(config) =>
+        val result = generate(
```

Comment:
nitpick: (blame it on my love for pure FP). Can we make `generate` return an `Either` instead of throwing exceptions. Then the `main` can deterministically have an exit value under all circumstances. Currently when `generate` throws we don't have an exit value and the caller will have to catch the exception.

```scala
type Error = String
def generate(..): Either[Error, String] = //..
```

and we return a `Left` for error with detailed messages and a `Right` for the actual value.

Again: Just a suggestion - feel free to ignore

Reply:
encapsulated the call site into a `Try` with a nicer error handling.

A couple of considerations:
 - this is, of course, still half baked, the `Main` is probably going to be removed at the end of the process
 - the `generate` method might be called by Java code and I want to keep the API as simple and agnostic as possible